### PR TITLE
refactor: nexus data base model

### PR DESCRIPTION
**Pull request by bengeos**
# Description

### Issue

Refactoring Database models by removing nexus & nexusUsers table

[Link to Basecamp Todo](https://3.basecamp.com/3105655/projects)

### Solution

Nexus, NexusUsers tables removed. nexus_id column removed from resources, channels

### CODE REVIEW

File: apps/api-nexus/db/schema.prisma [hunk 1]
```diff
@@ -145,27 +168,42 @@ enum BatchStatus {
...
+    createdAt DateTime        @default(now())
+    updatedAt DateTime        @updatedAt
 }
 
-model UserRole {
-    id     String @id @default(uuid())
-    userId String @unique
-    roles  Role[]
-
-    @@index(userId)
```
Comment (tataihono) [inline apps/api-nexus/db/schema.prisma:170]: this removal needs to be reverted

File: apps/api-nexus/db/schema.prisma [hunk 2]
```diff
@@ -145,27 +168,42 @@ enum BatchStatus {
...
+
+enum BatchTaskStatus {
+    pending
+    processing
+    completed
+    failed
 }
 
-enum Role {
-    publisher
```
Comment (tataihono) [inline apps/api-nexus/db/schema.prisma:162]: this removal needs to be reverted

File: apps/api-nexus/db/schema.prisma [hunk 3]
```diff
@@ -145,27 +168,42 @@ enum BatchStatus {
...
-    isCompleted Boolean  @default(false)
-    error       String?
-    percent     Float    @default(0.0)
-    batch       Batch    @relation(fields: [batchId], references: [id])
-    resource    Resource @relation(fields: [resourceId], references: [id])
-
-    @@unique([batchId, resourceId])
+model Batch {
+    id             String      @id @default(uuid())
+    nexusId        String
```
Comment (tataihono) [inline apps/api-nexus/db/schema.prisma:173]: are we removing the nexus object?

File: apps/api-nexus/db/schema.prisma [hunk 4]
```diff
@@ -103,34 +79,81 @@ enum ResourceStatus {
...
+    resourceLocalizationSource ResourceLocalizationSource?
+}
+
+model ResourceLocalizationSource {
+    id                                String               @id @default(uuid())
+    reourceLocalizationId             String               @unique
+    // CAPTION
+    captionGoogleDriveId              String?
+    captionGoogleDriveRefreshToken    String?
+    captionCloudFlareId               String?
```
Comment (tataihono) [inline apps/api-nexus/db/schema.prisma:149]: can we drop cloudflare?

File: apps/api-nexus/db/schema.prisma [hunk 5]
```diff
@@ -103,34 +79,81 @@ enum ResourceStatus {
...
+    customThumbnail       String?
+    playlistId            String?
+    isMadeForKids         Boolean                @default(false)
+    mediaComponentId      String?
+    notifySubscribers     Boolean                @default(false)
+    status                ResourceStatus         @default(published)
+    privacy               PrivacyStatus          @default(private)
+    resourceLocalizations ResourceLocalization[]
+    resourceChannels      ResourceChannel[]
+    nexus                 Nexus                  @relation(fields: [nexusId], references: [id])
```
Comment (tataihono) [inline apps/api-nexus/db/schema.prisma:97]: is it likely that resources are going to be connected to multiple channels?

File: apps/api-nexus/src/app/lib/bullMQ/consumers/batch.job.worker.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      async (progress) => {
+        progress = 30 + progress / 4
+        await job.progress(progress)
+        return await Promise.resolve()
+      }
+    )
+
+    console.log('resource', job?.data?.resource)
+
+    // UPLOAD VIDEO
```
Comment (tataihono) [inline apps/api-nexus/src/app/lib/bullMQ/consumers/batch.job.worker.ts:50]: drop all commented code. it can be recovered via git.

**Review by tataihono — CHANGES_REQUESTED** [review]

**Review by tataihono — CHANGES_REQUESTED** [review]
@bengeos can we also merge the channelYoutube and Channel models?

File: apps/api-nexus/db/schema.prisma [hunk 7]
```diff
@@ -8,91 +8,39 @@ datasource db {
...
-    status    NexusStatus     @default(published)
-    createdAt DateTime        @default(now())
-    deletedAt DateTime?
-    Batch     Batch[]
+    id               String            @id @default(uuid())
+    name             String
+    platform         String
+    connected        Boolean           @default(false)
+    youtube          ChannelYoutube?
+    status           ChannelStatus     @default(created)
```
Comment (tataihono) [inline apps/api-nexus/db/schema.prisma:23]: swap for publishedAt?

File: apps/api-nexus/db/schema.prisma [hunk 8]
```diff
@@ -145,17 +138,42 @@ enum BatchStatus {
...
+    batch     Batch           @relation(fields: [batchId], references: [id])
+    createdAt DateTime        @default(now())
+    updatedAt DateTime        @updatedAt
+}
 
-    @@unique([batchId, resourceId])
+model GoogleAccessToken {
+    id           String @id @default(uuid())
+    refreshToken String
 }
```
Comment (tataihono) [inline apps/api-nexus/db/schema.prisma:177]: is this needed anymore?

File: apps/api-nexus/db/seed.ts [hunk 9]
```diff
@@ -1,7 +1,5 @@
-import { jfNexus } from './seeds/jfNexus'
-
 async function main(): Promise<void> {
```
Comment (tataihono) [inline apps/api-nexus/db/seed.ts:1]: is this file needed anymore?

File: apps/api-nexus/src/app/lib/bullMQ/bullMQ.module.ts [hunk 10]
```diff
@@ -22,19 +22,16 @@ import { NexusJobListener } from './listener/jobListener'
...
   ],
   providers: [
     BullMQService,
-    NexusBucketProcessor,
-    NexusYoutubeProcessor,
-    GoogleDriveService,
+    BatchJobWorker,
     GoogleOAuthService,
     GoogleSheetsService,
+    GoogleDriveService,
```
Comment (tataihono) [inline apps/api-nexus/src/app/lib/bullMQ/bullMQ.module.ts:32]: did we decide to drop google drive?

File: apps/api-nexus/src/app/lib/bullMQ/bullMQ.service.ts [hunk 11]
```diff
@@ -46,76 +72,183 @@ export interface UploadToBucketToYoutube {
...
+            channelId: channel.id
+          }
         }
       })
     })
 
-    await this.createBatchJob(batch.id)
+    await this.createUploadResourceBatchJob(batch.id, channel)
 
     return batch as unknown as Batch
```
Comment (tataihono) [inline apps/api-nexus/src/app/lib/bullMQ/bullMQ.service.ts:172]: please don't type as unknown. batch needs to have the appropriate type you need to return.

File: apps/api-nexus/src/app/lib/bullMQ/bullMQ.service.ts [hunk 12]
```diff
@@ -46,76 +72,183 @@ export interface UploadToBucketToYoutube {
...
+              loc.resourceLocalizationSource?.captionGoogleDriveRefreshToken ??
+              ''
+          }
+        })
+      }
+      await this.batchJobQueue.addBulk([
+        { data: job, name: 'processResourceUpdate' }
+      ])
+    }
+    return batch as unknown as Batch
```
Comment (tataihono) [inline apps/api-nexus/src/app/lib/bullMQ/bullMQ.service.ts:252]: same for here.

File: apps/api-nexus/src/app/lib/bullMQ/consumers/batch.job.worker.ts [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+    const bucketFile = await this.bucketService.uploadFile(
+      videoFilePath,
+      process.env.BUCKET_NAME ?? 'bucket-name',
+      async (progress) => {
+        progress = 30 + progress / 4
+        await job.progress(progress)
+        return await Promise.resolve()
+      }
+    )
+
```
Comment (tataihono) [inline apps/api-nexus/src/app/lib/bullMQ/consumers/batch.job.worker.ts:218]: uploading to cloudflare is not necessary.

File: apps/api-nexus/src/app/lib/bullMQ/consumers/batch.job.worker.ts [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+    // UPLOAD VIDEO
+    console.log('UPLOAD TO YOUTUBE NOW', videoFilePath)
```
Comment (tataihono) [inline apps/api-nexus/src/app/lib/bullMQ/consumers/batch.job.worker.ts:220]: drop this log.

File: apps/api-nexus/src/app/lib/bullMQ/listener/jobListener.ts [hunk 15]
```diff
@@ -3,14 +3,13 @@ import { InjectQueue } from '@nestjs/bull'
 import { Injectable, OnModuleInit } from '@nestjs/common'
 import { Job, Queue } from 'bull'
 
-import { PrismaService } from '../../prisma.service'
-import { UploadToBucketToYoutube } from '../bullMQ.service'
+// import { PrismaService } from '../../prisma.service'
```
Comment (tataihono) [inline apps/api-nexus/src/app/lib/bullMQ/listener/jobListener.ts:6]: drop commented code.

File: apps/api-nexus/src/app/lib/google/drive.service.ts [hunk 16]
```diff
@@ -23,26 +19,9 @@ type FileResponse = Pick<
...
-  title?: string
-  description?: string
-  keywords?: string
-  category?: string
-  privacy?: string
-  spoken_language?: string
-}
-
 @Injectable()
 export class GoogleDriveService {
```
Comment (tataihono) [inline apps/api-nexus/src/app/lib/google/drive.service.ts:23]: this service needs to be dropped right?

File: apps/api-nexus/src/app/lib/google/sheets.service.ts [hunk 17]
```diff
@@ -1,8 +1,62 @@
+import { drive_v3 } from '@googleapis/drive'
```
Comment (tataihono) [inline apps/api-nexus/src/app/lib/google/sheets.service.ts:1]: we were swapping this out for csv right?

**Review by tataihono — CHANGES_REQUESTED** [review]

File: apps/api-nexus/db/schema.prisma (hunk 7 above)
Comment (bengeos) [reply apps/api-nexus/db/schema.prisma:23]: this change is introduced @ https://github.com/JesusFilm/core/pull/2771

File: apps/api-nexus/db/schema.prisma (hunk 8 above)
Comment (bengeos) [reply apps/api-nexus/db/schema.prisma:177]: this change is introduced @ https://github.com/JesusFilm/core/pull/2771

File: apps/api-nexus/src/app/lib/bullMQ/bullMQ.module.ts (hunk 10 above)
Comment (bengeos) [reply apps/api-nexus/src/app/lib/bullMQ/bullMQ.module.ts:32]: For time being we are still using the google drive library to download the public drive links. We will completely remove this on the next iteration

File: apps/api-nexus/db/seed.ts (hunk 9 above)
Comment (bengeos) [reply apps/api-nexus/db/seed.ts:1]: this is already done @ https://github.com/JesusFilm/core/pull/2875

**Pull request merged**